pkg/controller/rollout/rollout_status.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2022 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rollout
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/openkruise/rollouts/api/v1alpha1"
+	"github.com/openkruise/rollouts/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func (r *RolloutReconciler) updateRolloutStatus(rollout *v1alpha1.Rollout) (retry bool, err error) {
+	newStatus := *rollout.Status.DeepCopy()
+	newStatus.ObservedGeneration = rollout.GetGeneration()
+	defer func() {
+		err = r.updateRolloutStatusInternal(rollout, newStatus)
```

Comment:
why update status here? there is also updateRolloutStatusInternal in reconcileRolloutProgressing and reconcileRolloutTerminating, update status twice in a single reconciliation loop is not recommended